File: app/src/main/java/org/schabi/newpipe/settings/preferencesearch/PreferenceSearchResultHighlighter.java
package org.schabi.newpipe.settings.preferencesearch;

import android.content.res.Resources;
import android.content.res.TypedArray;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.RippleDrawable;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.util.TypedValue;

import androidx.appcompat.content.res.AppCompatResources;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceGroup;
import androidx.recyclerview.widget.RecyclerView;

import org.schabi.newpipe.R;


public final class PreferenceSearchResultHighlighter {
    private static final String TAG = "PrefSearchResHighlter";

    private PreferenceSearchResultHighlighter() {
    }

    /**
     * Highlight the specified preference.
     * <br/>
     * Note: This function is Thread independent (can be called from outside of the main thread).
     *
     * @param item The item to highlight
     * @param prefsFragment The fragment where the items is located on
     */
    public static void highlight(
            final PreferenceSearchItem item,
            final PreferenceFragmentCompat prefsFragment
    ) {
        new Handler(Looper.getMainLooper()).post(() -> doHighlight(item, prefsFragment));
    }

    private static void doHighlight(
            final PreferenceSearchItem item,
            final PreferenceFragmentCompat prefsFragment
    ) {
        final Preference prefResult = prefsFragment.findPreference(item.getKey());

        if (prefResult == null) {
            Log.w(TAG, "Preference '" + item.getKey() + "' not found on '" + prefsFragment + "'");
            return;
        }

        final RecyclerView recyclerView = prefsFragment.getListView();
        final RecyclerView.Adapter<?> adapter = recyclerView.getAdapter();
        if (adapter instanceof PreferenceGroup.PreferencePositionCallback) {
            final int position = ((PreferenceGroup.PreferencePositionCallback) adapter)
                    .getPreferenceAdapterPosition(prefResult);
            if (position != RecyclerView.NO_POSITION) {
                recyclerView.scrollToPosition(position);
                recyclerView.postDelayed(() -> {
                    final RecyclerView.ViewHolder holder =
                            recyclerView.findViewHolderForAdapterPosition(position);
                    if (holder != null) {
                        final Drawable background = holder.itemView.getBackground();
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP
                                && background instanceof RippleDrawable) {
                            showRippleAnimation((RippleDrawable) background);
                            return;
                        }
                    }
                    highlightFallback(prefsFragment, prefResult);
                }, 200);
                return;
            }
        }
        highlightFallback(prefsFragment, prefResult);
    }

    /**
     * Alternative highlighting (shows an → arrow in front of the setting)if ripple does not work.
     *
     * @param prefsFragment
     * @param prefResult
     */
    private static void highlightFallback(
            final PreferenceFragmentCompat prefsFragment,
            final Preference prefResult
    ) {
        // Get primary color from text for highlight icon
        final TypedValue typedValue = new TypedValue();
        final Resources.Theme theme = prefsFragment.getActivity().getTheme();
        theme.resolveAttribute(android.R.attr.textColorPrimary, typedValue, true);
        final TypedArray arr = prefsFragment.getActivity()
                .obtainStyledAttributes(
                        typedValue.data,
                        new int[]{android.R.attr.textColorPrimary});
        final int color = arr.getColor(0, 0xffE53935);
        arr.recycle();

        // Show highlight icon
        final Drawable oldIcon = prefResult.getIcon();
        final boolean oldSpaceReserved = prefResult.isIconSpaceReserved();
        final Drawable highlightIcon =
                AppCompatResources.getDrawable(
                        prefsFragment.requireContext(),
                        R.drawable.ic_play_arrow);
        highlightIcon.setColorFilter(new PorterDuffColorFilter(color, PorterDuff.Mode.SRC_IN));
        prefResult.setIcon(highlightIcon);

        prefsFragment.scrollToPreference(prefResult);

        new Handler(Looper.getMainLooper()).postDelayed(() -> {
            prefResult.setIcon(oldIcon);
            prefResult.setIconSpaceReserved(oldSpaceReserved);
        }, 1000);
    }

    private static void showRippleAnimation(final RippleDrawable rippleDrawable) {
        rippleDrawable.setState(
                new int[]{android.R.attr.state_pressed, android.R.attr.state_enabled});
        new Handler(Looper.getMainLooper())
                .postDelayed(() -> rippleDrawable.setState(new int[]{}), 1000);
    }
}

File: app/src/main/java/org/schabi/newpipe/settings/preferencesearch/PreferenceSearchResultHighlighter.kt
package org.schabi.newpipe.settings.preferencesearch

import android.content.res.Resources.Theme
import android.content.res.TypedArray
import android.graphics.PorterDuff
import android.graphics.PorterDuffColorFilter
import android.graphics.drawable.Drawable
import android.graphics.drawable.RippleDrawable
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.util.TypedValue
import androidx.appcompat.content.res.AppCompatResources
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceGroup.PreferencePositionCallback
import androidx.recyclerview.widget.RecyclerView
import org.schabi.newpipe.R

object PreferenceSearchResultHighlighter {
    private val TAG: String = "PrefSearchResHighlter"

    /**
     * Highlight the specified preference.
     * <br></br>
     * Note: This function is Thread independent (can be called from outside of the main thread).
     *
     * @param item The item to highlight
     * @param prefsFragment The fragment where the items is located on
     */
    fun highlight(
            item: PreferenceSearchItem,
            prefsFragment: PreferenceFragmentCompat
    ) {
        Handler(Looper.getMainLooper()).post(Runnable({ doHighlight(item, prefsFragment) }))
    }

    private fun doHighlight(
            item: PreferenceSearchItem,
            prefsFragment: PreferenceFragmentCompat
    ) {
        val prefResult: Preference? = prefsFragment.findPreference(item.getKey())
        if (prefResult == null) {
            Log.w(TAG, "Preference '" + item.getKey() + "' not found on '" + prefsFragment + "'")
            return
        }
        val recyclerView: RecyclerView = prefsFragment.getListView()
        val adapter: RecyclerView.Adapter<*>? = recyclerView.getAdapter()
        if (adapter is PreferencePositionCallback) {
            val position: Int = (adapter as PreferencePositionCallback)
                    .getPreferenceAdapterPosition(prefResult)
            if (position != RecyclerView.NO_POSITION) {
                recyclerView.scrollToPosition(position)
                recyclerView.postDelayed(Runnable({
                    val holder: RecyclerView.ViewHolder? = recyclerView.findViewHolderForAdapterPosition(position)
                    if (holder != null) {
                        val background: Drawable = holder.itemView.getBackground()
                        if (background is RippleDrawable) {
                            showRippleAnimation(background)
                            return@postDelayed
                        }
                    }
                    highlightFallback(prefsFragment, prefResult)
                }), 200)
                return
            }
        }
        highlightFallback(prefsFragment, prefResult)
    }

    /**
     * Alternative highlighting (shows an → arrow in front of the setting)if ripple does not work.
     *
     * @param prefsFragment
     * @param prefResult
     */
    private fun highlightFallback(
            prefsFragment: PreferenceFragmentCompat,
            prefResult: Preference
    ) {
        // Get primary color from text for highlight icon
        val typedValue: TypedValue = TypedValue()
        val theme: Theme = prefsFragment.getActivity()!!.getTheme()
        theme.resolveAttribute(android.R.attr.textColorPrimary, typedValue, true)
        val arr: TypedArray = prefsFragment.getActivity()
                .obtainStyledAttributes(
                        typedValue.data, intArrayOf(android.R.attr.textColorPrimary))
        val color: Int = arr.getColor(0, -0x1ac6cb)
        arr.recycle()

        // Show highlight icon
        val oldIcon: Drawable? = prefResult.getIcon()
        val oldSpaceReserved: Boolean = prefResult.isIconSpaceReserved()
        val highlightIcon: Drawable? = AppCompatResources.getDrawable(
                prefsFragment.requireContext(),
                R.drawable.ic_play_arrow)
        highlightIcon!!.setColorFilter(PorterDuffColorFilter(color, PorterDuff.Mode.SRC_IN))
        prefResult.setIcon(highlightIcon)
        prefsFragment.scrollToPreference(prefResult)
        Handler(Looper.getMainLooper()).postDelayed(Runnable({
            prefResult.setIcon(oldIcon)
            prefResult.setIconSpaceReserved(oldSpaceReserved)
        }), 1000)
    }

    private fun showRippleAnimation(rippleDrawable: RippleDrawable) {
        rippleDrawable.setState(intArrayOf(android.R.attr.state_pressed, android.R.attr.state_enabled))
        Handler(Looper.getMainLooper())
                .postDelayed(Runnable({ rippleDrawable.setState(intArrayOf()) }), 1000)
    }
}
